### PR TITLE
Add DeepSpeed to LightningStableDiffusion model

### PR DESCRIPTION
**Pull request by tchaton**
This PR extracts the DeepSpeed Stable Diffusion Inference Optimization Policy applied over Diffusers models when doing the following.

```python
pipe = diffusers.StableDiffusionPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5",
    use_auth_token=hf_auth_key,
    torch_dtype=torch.float16,
    revision="fp16")

print(pipe)
pipe = deepspeed.init_inference(pipe.to("cuda"), dtype=torch.float16)
```

This PR introduces the following:
* A generic Cuda Graph Replay with dynamic batch sizes and custom ones for Unet, VAE, ClipEncoder.
* A custom policy for Stable Diffusion Unet to replace CrossAttention, TransformerBlock by DeepSpeed optimized one.
* Perform inference in float16
* Add benchmarks scripts to compare HF and our implementation. CF Table: https://www.notion.so/lightningai/Optimising-Stable-Diffusion-5a29e10606884c28b15616108b96236c#5309c54655e34e21a62025db861cf5f1
* Enable to easily switch between 3 schedulers and default to the SOTA: https://github.com/LuChengTHU/dpm-solver

### CODE REVIEW

File: ldm/lightning.py [hunk 1]
```diff
@@ -160,12 +124,12 @@ def __init__(
...
         self.model.load_state_dict(sd, strict=False)
-        
+
         # Update Unet for inference
         # Currently waiting for https://github.com/pytorch/pytorch/issues/91302
-        # self.model.model = ReplayCudaGraphUnet(self.model.model)
+        self.model.model = ReplayCudaGraphUnet(self.model.model)
 
-        self.to(device)
+        self.to(device, dtype=torch.float16)
```
Comment (lantiga) [inline ldm/lightning.py:134]: precision scope is conditioned on whether precision == 16, but here we are hard-coding it
would it make sense at all to keep it optional?

File: ldm/deepspeed_replace.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
+import torch
```
Comment (hhsecond) [inline ldm/deepspeed_replace.py:5]: should we keep some licensing information on the top?

Comment (lantiga) [reply ldm/deepspeed_replace.py:5]: +1 for me, the code draws heavily from the deepspeed repo

File: scripts/txt2img.py [hunk 3]
```diff
@@ -276,9 +266,14 @@ def main(opt):
             # to image
             grid = 255. * rearrange(grid, 'c h w -> h w c').cpu().numpy()
             grid = Image.fromarray(grid.astype(np.uint8))
-            grid = put_watermark(grid, wm_encoder)
             grid.save(os.path.join(outpath, f'grid-{grid_count:04}.png'))
             grid_count += 1
+    
+    end_event.record()
```
Comment (lantiga) [inline scripts/txt2img.py:272]: does this ensure `synchronized` is called again?

**Review by lantiga — APPROVED** [review]
Looks great, learned a lot :-) I added a couple of comments

**Review by hhsecond — APPROVED** [review]

**Pull request merged**